src/ADSR.cpp

```diff
@@ -0,0 +1,359 @@
+					timeInCurrentStage = 0.f;
+					if (gateHeld) {
+						stage = STAGE_SUSTAIN;
+					}
+					else {
+						stage = STAGE_RELEASE;
+						releaseValue = env;
+					}
+				}
+				break;
+			}
+			case STAGE_OFF:
+			case STAGE_RELEASE:
+			case STAGE_SUSTAIN: {
+				break;
+			}
+		}
+
+		if (!gateHeld) {
+			if (stage == STAGE_DECAY || stage == STAGE_SUSTAIN) {
+				timeInCurrentStage = 0.f;
+				stage = STAGE_RELEASE;
+				releaseValue = env;
+			}
+			else if (stage == STAGE_RELEASE) {
+				if (timeInCurrentStage >= releaseTime) {
+					stage = STAGE_OFF;
+					timeInCurrentStage = 0.f;
+				}
+			}
+		}
+	}
+
+	void evolveEnvelope(const float& sampleTime) {
+		switch (stage) {
+			case  STAGE_OFF: {
+				env = 0.0f;
+				break;
+			}
+			case STAGE_ATTACK: {
+				timeInCurrentStage += sampleTime;
+				env = std::min(timeInCurrentStage / attackTime, 1.f);
+				env = std::pow(env, attackShape);
+				break;
+			}
+			case STAGE_DECAY: {
+				timeInCurrentStage += sampleTime;
+				env = std::pow(1.f - std::min(1.f, timeInCurrentStage / decayTime), decayShape);
+				env = sustainLevel + (1.f - sustainLevel) * env;
+				break;
+			}
+			case STAGE_SUSTAIN: {
+				env = sustainLevel;
+				break;
+			}
+			case STAGE_RELEASE: {
+				timeInCurrentStage += sampleTime;
+				env = std::min(1.0f, timeInCurrentStage / releaseTime);
+				env = releaseValue * std::pow(1.0f - env, releaseShape);
+				break;
+			}
+		}
+	}
+
+	void process(const float& sampleTime, const bool& gateHeld, const bool& triggerMode) {
+
+		if (triggerMode) {
+			processTransitionsTriggerMode(gateHeld);
+		}
+		else {
+			processTransitionsGateMode(gateHeld);
+		}
+
+		evolveEnvelope(sampleTime);
+	}
+};
+
+struct ADSR : Module {
+	enum ParamIds {
+		TRIGG_GATE_TOGGLE_PARAM,
+		MANUAL_TRIGGER_PARAM,
+		SHAPE_PARAM,
+		ATTACK_PARAM,
+		DECAY_PARAM,
+		SUSTAIN_PARAM,
+		RELEASE_PARAM,
+		NUM_PARAMS
+	};
+	enum InputIds {
+		TRIGGER_INPUT,
+		CV_ATTACK_INPUT,
+		CV_DECAY_INPUT,
+		CV_SUSTAIN_INPUT,
+		CV_RELEASE_INPUT,
+		NUM_INPUTS
+	};
+	enum OutputIds {
+		OUT_OUTPUT,
+		STAGE_ATTACK_OUTPUT,
+		STAGE_DECAY_OUTPUT,
+		STAGE_SUSTAIN_OUTPUT,
+		STAGE_RELEASE_OUTPUT,
+		NUM_OUTPUTS
+	};
+	enum LightIds {
+		LED_LIGHT,
+		LED_ATTACK_LIGHT,
+		LED_DECAY_LIGHT,
+		LED_SUSTAIN_LIGHT,
+		LED_RELEASE_LIGHT,
+		NUM_LIGHTS
+	};
+	enum EnvelopeMode {
+		GATE_MODE,
+		TRIGGER_MODE
+	};
+
+	BefacoADSREnvelope envelope;
+	dsp::SchmittTrigger gateTrigger;
+	dsp::ClockDivider cvDivider;
+	float shape;
+
+	static constexpr float minStageTime = 0.003f;  // in seconds
+	static constexpr float maxStageTime = 10.f;  // in seconds
+
+	// given a value from the slider and/or cv (rescaled to range 0 to 1), transform into the appropriate time in seconds
+	static float convertCVToTimeInSeconds(float cv) {
+		float cv2 = cv * cv;
+		// according to hardware, slider appears to respond roughly as a quartic
+		return minStageTime + (maxStageTime - minStageTime) * cv2 * cv2;
```

Comment:
Quartic with a nonzero min/max can be very close to exponential scaling, which is available in `configParam()` arguments. Have you considered switching to that?

Reply:
Good point, the follow works pretty well so will update to that: 

https://www.desmos.com/calculator/z2vbqdxtut (annoying no log scale!)

Reply:
Voltage-to-time should actually always be `2^x` scale to automatically follow 1V/oct. https://vcvrack.com/manual/VoltageStandards
The reason this is important is because you can't change it later. It'll break people's patches.

Reply:
First of all I agree it's important to get it right now, so I appreciate the input!

The value I get there is just the calibration relation between the slider and the behaviour I recorded on the scope (same for CV). It does affect the playability (e.g. when I got the relation wrong to start it was hard to dial in an envelope I wanted). I think it would be hard to get something sensible in the format `a + b 2^x` that's close to what the hardware does.

My hardware ADSR doesn't track 1V/Oct nor would I expect it to (same for e.g. Rampage I think?). I can certainly see the value for VCOs / LFOs to adhere to the standard. 

I guess this comes down to sticking to hardware "realism" vs "consistent expected behaviour". My personal preference is for the former in this case. Were I creating a new module from scratch, the latter.

Reply:
Oh right, I forgot you're using the `convertCVToTimeInSeconds()` for parameters. The name threw me off but for parameters, using some high base (like 200 instead of 2) will fit `x^4` closely.
Your current code looks fine then.